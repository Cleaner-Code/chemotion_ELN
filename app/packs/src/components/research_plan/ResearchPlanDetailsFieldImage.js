import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Dropzone from 'react-dropzone';
import { Button, FormControl, FormGroup, InputGroup } from 'react-bootstrap';
import ResearchPlansFetcher from '../fetchers/ResearchPlansFetcher';
import ImageEditModal from './ImageEditModal';
import logger from 'redux-logger';

export default class ResearchPlanDetailsFieldImage extends Component {

  constructor(props){
    super(props);
    this.state={imageEditModalShown:false}
 

  }

  handleDrop(files) {
    const { field, onChange } = this.props;
    const imageFile = files[0];
    const replace = field.value.public_name;

    // upload new image
    ResearchPlansFetcher.updateImageFile(imageFile, replace).then((value) => {
      // update research plan
      onChange(value, field.id);
    });
  }

  handleResizeChange(event) {
    const { field, onChange } = this.props;
    field.value.zoom = event.target.value;
    onChange(field.value, field.id);
  }

  renderEdit() {

    const { field } = this.props;
    let content;
    if (field.value.public_name) {
      const src = `/images/research_plans/${field.value.public_name}`;
      const style = (field.value.zoom == null || typeof field.value.zoom === 'undefined'
      || field.value.width === '') ? { width: 'unset' } : { width: `${field.value.zoom}%` };
      content = (
        <div className="image-container">
          <img id="researchPlanImageID" style={style} src={src} alt={field.value.file_name} />
        </div>
      );
    } else {
      content = <p>Drop Files, or Click to Select.</p>;
    }
    return (
      <div>
       <ImageEditModal
          imageName={field.value.public_name}
          isShow={this.state.imageEditModalShown}
          handleSave={(f)=>{this.handleDrop(f);this.setState({imageEditModalShown:false})}}
          handleOnClose={()=>{this.setState({imageEditModalShown:false})}}
       />
     
       
        <FormGroup style={{ width: '30%' }}>
          <InputGroup>
            <InputGroup.Addon>Zoom</InputGroup.Addon>
            <FormControl
              type="number"
              max="100"
              min="1"
              placeholder="image zoom"
              defaultValue={field.value.zoom}
              onChange={event => this.handleResizeChange(event)}
            />
            <InputGroup.Addon>%</InputGroup.Addon>
            <Button
            onClick={()=>{this.setState({imageEditModalShown:true})}}>Edit</Button>
          </InputGroup>
        
        </FormGroup>       
      
        <Dropzone
          accept="image/*"
          multiple={false}
          onDrop={files => this.handleDrop(files)}
          className="dropzone"
        >
          {content}
        </Dropzone>
      </div>
    );
  }

  renderStatic() {
    const { field } = this.props;
    if (typeof (field.value.public_name) === 'undefined'
    || field.value.public_name === null) {
      return (
        <div />
      );
    }
    const src = `/images/research_plans/${field.value.public_name}`;
    const style = (field.value.zoom == null || typeof field.value.zoom === 'undefined'
    || field.value.width === '') ? { width: 'unset' } : { width: `${field.value.zoom}%` };

    return (
      <div className="image-container">
        <img style={style} src={src} alt={field.value.file_name} />
      </div>
    );
  }

  render() {
    if (this.props.edit) {
      return this.renderEdit();
    }
    return this.renderStatic();
  }
}

ResearchPlanDetailsFieldImage.propTypes = {
  field: PropTypes.object,
  index: PropTypes.number,
  disabled: PropTypes.bool,
  onChange: PropTypes.func,
  edit: PropTypes.bool,
};
