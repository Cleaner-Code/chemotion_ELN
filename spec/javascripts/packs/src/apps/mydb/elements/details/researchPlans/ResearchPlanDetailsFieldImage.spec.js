import React from 'react';
import expect from 'expect';
import Enzyme, { shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import ResearchPlanDetailsFieldImage from 'src/apps/mydb/elements/details/researchPlans/researchPlanTab/ResearchPlanDetailsFieldImage';
import ResearchPlan from 'src/models/ResearchPlan';
import AttachmentFetcher from 'src/fetchers/AttachmentFetcher';
import sinon from 'sinon';
import toMatchSnapshot from 'expect-mocha-snapshot';
import { describe, it } from 'mocha';
import ResearchPlanFactory from 'factories/ResearchPlanFactory';

Enzyme.configure({ adapter: new Adapter() });

describe('ResearchPlanDetailsFieldImage', () => {
  expect.extend({ toMatchSnapshot });

  describe('.render()', () => {
    describe('no image choosen in view mode', async () => {
      it('no img tag should be rendered', async function () {
        const researchPlan = await ResearchPlanFactory.build('empty', {}, { reset: true });

        const wrapper = shallow(<ResearchPlanDetailsFieldImage field={researchPlan} />);

        expect(wrapper.html()).toMatchSnapshot(this);
      });
    });
    describe('temporary image in blob in view mode', async () => {
      it('img tag present, src should be starting with blob', async function () {
        const researchPlan = await ResearchPlanFactory.build('empty', {}, { reset: true });

        researchPlan.value = {};
        researchPlan.value.public_name = 'blob://...';
        researchPlan.value.file_name = 'myFile.png';

        const wrapper = shallow(<ResearchPlanDetailsFieldImage field={researchPlan} />);
        expect(wrapper.html()).toMatchSnapshot(this);
      });
    });
    describe('deprecated image in view mode', async () => {
      it('img tag present, src should be starting with absolute path /images/research_plans', async function () {
        const rp = ResearchPlan.buildEmpty();

        rp.value = {};
        rp.value.public_name = 'xxx.png';
        rp.value.file_name = 'xxx.png';

        const wrapper = shallow(<ResearchPlanDetailsFieldImage field={rp} />);
        expect(wrapper.html()).toMatchSnapshot(this);
      });
    });
    describe('image from attachment on server in view mode', async () => {
      it('img tag present, src should be the scrToImage from the stubed method', async function () {
        sinon
          .stub(AttachmentFetcher, 'fetchImageAttachmentByIdentifier')
          .callsFake((src) => { const result = []; result.data = src.identifier; return Promise.resolve(result); });

        const rp = ResearchPlan.buildEmpty();

        rp.value = {};
        rp.value.public_name = 'scrToImageOnServer';
        rp.value.file_name = 'fileName';

        const wrapper = await shallow(<ResearchPlanDetailsFieldImage field={rp} />);

        expect(wrapper.html()).toMatchSnapshot(this);
      });
    });
  });
});
