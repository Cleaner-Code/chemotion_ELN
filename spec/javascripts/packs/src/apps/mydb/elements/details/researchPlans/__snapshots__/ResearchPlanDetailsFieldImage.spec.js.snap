// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResearchPlanDetailsFieldImage .render() deprecated image in view mode img tag present, src should be starting with absolute path /images/research_plans 1`] = `"<div class=\\"image-container\\"><img style=\\"width:unset\\" src=\\"/images/research_plans/xxx.png\\" alt=\\"xxx.png\\"/></div>"`;

exports[`ResearchPlanDetailsFieldImage .render() image from attachment on server in view mode img tag present, src should be the scrToImage from the stubed method 1`] = `"<div class=\\"image-container\\"><img style=\\"width:unset\\" src=\\"scrToImageOnServer\\" alt=\\"fileName\\"/></div>"`;

exports[`ResearchPlanDetailsFieldImage .render() no image choosen in view mode no img tag should be rendered 1`] = `"<div></div>"`;

exports[`ResearchPlanDetailsFieldImage .render() temporary image in blob in view mode img tag present, src should be starting with blob 1`] = `"<div class=\\"image-container\\"><img style=\\"width:unset\\" src=\\"blob://...\\" alt=\\"myFile.png\\"/></div>"`;
